### PR TITLE
[Merged by Bors] - feat(Combinatorics/SimpleGraph): the reverse of a cycle is a cycle

### PR DESCRIPTION
**Pull request by pimotte**
Added the result that a reverse cycle is a cycle. 
Includes supporting lemmas concerning lists. In preparation for Tutte's theorem.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

[Thread on Tuttes](https://leanprover.zulipchat.com/#narrow/channel/287929-mathlib4/topic/Tutte's.20theorem)

### CODE REVIEW

File: Mathlib/Combinatorics/SimpleGraph/Path.lean [hunk 1]
```diff
@@ -249,6 +249,10 @@ theorem cons_isCycle_iff {u v : V} (p : G.Walk v u) (h : G.Adj u v) :
   have : p.support.Nodup → p.edges.Nodup := edges_nodup_of_support_nodup
   tauto
 
+lemma IsCycle.reverse {p : G.Walk u u} (h : p.IsCycle) : p.reverse.IsCycle := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Path.lean:252]:
```suggestion
protected lemma IsCycle.reverse {p : G.Walk u u} (h : p.IsCycle) : p.reverse.IsCycle := by
```

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 2]
```diff
@@ -557,6 +557,20 @@ theorem subset_support_append_right {V : Type u} {G : SimpleGraph V} {u v w : V}
   intro h
   simp +contextual only [mem_support_append_iff, or_true, imp_true_iff]
 
+lemma getVert_eq_support_get? {u v n} (p : G.Walk u v) (h2 : n ≤ p.length) :
+    p.getVert n = (p.support[n]?) := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:561]: Do you need the brackets here?
```suggestion
    p.getVert n = p.support[n]? := by
```

File: Mathlib/Combinatorics/SimpleGraph/Walk.lean [hunk 3]
```diff
@@ -739,6 +753,15 @@ theorem edges_nodup_of_support_nodup {u v : V} {p : G.Walk u v} (h : p.support.N
     simp only [edges_cons, support_cons, List.nodup_cons] at h ⊢
     exact ⟨fun h' => h.1 (fst_mem_support_of_mem_edges p' h'), ih h.2⟩
 
+@[simp]
+theorem reverse_support_tail_nodup {u : V} {p : G.Walk u u} :
+    p.reverse.support.tail.Nodup ↔ p.support.tail.Nodup := by
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Walk.lean:758]:
```suggestion
theorem nodup_tail_support_reverse {u : V} {p : G.Walk u u} :
    p.reverse.support.tail.Nodup ↔ p.support.tail.Nodup := by
```

File: Mathlib/Data/List/Nodup.lean [hunk 4]
```diff
@@ -227,6 +227,22 @@ theorem Nodup.filter (p : α → Bool) {l} : Nodup l → Nodup (filter p l) := b
 theorem nodup_reverse {l : List α} : Nodup (reverse l) ↔ Nodup l :=
   pairwise_reverse.trans <| by simp only [Nodup, Ne, eq_comm]
 
+@[simp]
+lemma nodup_tail_reverse (l : List α) (h : l[0]? = l.getLast?) :
+    Nodup l.reverse.tail ↔ Nodup l.tail := by
```
Comment (YaelDillies) [inline Mathlib/Data/List/Nodup.lean:232]: Do you think is a good simp lemma? I doubt the `h` condition will be found by simp

**Review by YaelDillies — COMMENTED** [review]
Thanks! Can we also have the iff `isCycle_reverse`?

File: Mathlib/Data/List/Nodup.lean (hunk 4 above)
Comment (pimotte) [reply Mathlib/Data/List/Nodup.lean]: Makes sense, removed it:)

File: Mathlib/Combinatorics/SimpleGraph/Path.lean [hunk 5]
```diff
@@ -249,6 +249,17 @@ theorem cons_isCycle_iff {u v : V} (p : G.Walk v u) (h : G.Adj u v) :
...
+protected lemma IsCycle.reverse {p : G.Walk u u} (h : p.IsCycle) : p.reverse.IsCycle := by
+  simp only [Walk.isCycle_def, nodup_tail_support_reverse] at h ⊢
+  exact ⟨h.1.reverse, fun h' ↦ h.2.1 (by simp_all [← Walk.length_eq_zero_iff]), h.2.2⟩
+
+@[simp]
+lemma isCycle_reverse {p : G.Walk u u} : p.reverse.IsCycle ↔ p.IsCycle := by
+  refine ⟨?_, fun h ↦ h.reverse⟩
+  intro h
+  rw [← reverse_reverse p]
+  exact h.reverse
```
Comment (YaelDillies) [inline Mathlib/Combinatorics/SimpleGraph/Path.lean:261]:
```suggestion
lemma isCycle_reverse {p : G.Walk u u} : p.reverse.IsCycle ↔ p.IsCycle where
  mp h := by simpa using h.reverse
  mpr := .reverse
```

**Review by YaelDillies — APPROVED** [review]
Thanks! 🚀

maintainer delegate

**Comment by j-loreaux** [comment]
bors d+

**Pull request closed without merging**